packages/core/src/2d/sprite/SpriteRenderer.ts

```diff
@@ -136,111 +203,59 @@ export class SpriteRenderer extends Renderer implements ICustomClone {
   }
 
   /**
    * @internal
    */
   _render(camera: Camera): void {
     const { sprite } = this;
-    if (!sprite) {
+    if (!sprite || !sprite.texture) {
       return;
     }
-    const { texture } = sprite;
-    if (!texture) {
-      return;
-    }
-
-    const { _positions } = this;
-    const { transform } = this.entity;
-
-    // Update sprite data.
-    sprite._updateMesh();
-
-    if (this._isWorldMatrixDirty.flag || this._spriteDirty.flag) {
-      const localPositions = sprite._positions;
-      const localVertexPos = SpriteRenderer._tempVec3;
-      const worldMatrix = transform.worldMatrix;
-      const { flipX, flipY } = this;
-
-      for (let i = 0, n = _positions.length; i < n; i++) {
-        const curVertexPos = localPositions[i];
-        localVertexPos.set(flipX ? -curVertexPos.x : curVertexPos.x, flipY ? -curVertexPos.y : curVertexPos.y, 0);
-        Vector3.transformToVec3(localVertexPos, worldMatrix, _positions[i]);
-      }
-
-      this._setDirtyFlagFalse(DirtyFlag.Flip);
-      this._isWorldMatrixDirty.flag = false;
-      this._spriteDirty.flag = false;
-      this._cacheFlipX = flipX;
-      this._cacheFlipY = flipY;
-    } else if (this._isContainDirtyFlag(DirtyFlag.Flip)) {
-      const { flipX, flipY } = this;
-      const flipXChange = this._cacheFlipX !== flipX;
-      const flipYChange = this._cacheFlipY !== flipY;
-
-      if (flipXChange || flipYChange) {
-        const { x, y } = transform.worldPosition;
-
-        for (let i = 0, n = _positions.length; i < n; i++) {
-          const curPos = _positions[i];
-
-          if (flipXChange) {
-            curPos.x = x * 2 - curPos.x;
-          }
-          if (flipYChange) {
-            curPos.y = y * 2 - curPos.y;
-          }
-        }
-      }
 
-      this._setDirtyFlagFalse(DirtyFlag.Flip);
-      this._cacheFlipX = flipX;
-      this._cacheFlipY = flipY;
+    // Update position.
+    if (this._transformChangeFlag.flag || this._dirtyFlag & DirtyFlag.Position) {
+      this._assembler.updatePositions(this);
+      this._dirtyFlag &= ~DirtyFlag.Position;
+      this._transformChangeFlag.flag = false;
     }
 
-    if (this._isContainDirtyFlag(DirtyFlag.MaskInteraction)) {
-      this._updateStencilState();
-      this._setDirtyFlagFalse(DirtyFlag.MaskInteraction);
+    // Update uv.
+    if (this._dirtyFlag & DirtyFlag.UV) {
+      this._assembler.updateUVs(this);
+      this._dirtyFlag &= ~DirtyFlag.UV;
     }
 
-    this.shaderData.setTexture(SpriteRenderer._textureProperty, texture);
-    const material = this.getMaterial();
-
-    const spriteElementPool = this._engine._spriteElementPool;
-    const spriteElement = spriteElementPool.getFromPool();
-    spriteElement.setValue(this, _positions, sprite._uv, sprite._triangles, this.color, material, camera);
+    // Push primitive.
+    const spriteElement = this._engine._spriteElementPool.getFromPool();
+    spriteElement.setValue(this, this._renderData, this.getMaterial());
```

Comment:
You already have `_renderData`, so `spriteElement` and `renderData` can merge  as `spriteRendererElement`,  you can delete `_spriteElementPool `

Reply:
It will be optimized in the next version.